src/meili/config.cc

```diff
@@ -0,0 +1,113 @@
+      throw std::invalid_argument(std::string("Expect '") + name +                                   \
+                                  "' to be nonnegative (got: " + std::to_string(value) + ")");       \
+  }
+
+namespace {
+template <typename T>
+inline void
+ReadParamOptional(T& value, const boost::property_tree::ptree& ptree, std::string const& name) {
+  if (auto item = ptree.get_optional<T>(name))
+    value = *item;
+}
+} // namespace
+
+namespace valhalla {
+namespace meili {
+
+Config::Config(const boost::property_tree::ptree& params) {
+  Read(params);
+}
+
+void Config::Read(const boost::property_tree::ptree& params) {
+  candidate_search.Read(params);
+  transition.Read(params);
+  emission.Read(params);
+  routing.Read(params);
+  customizable.Read(params);
+}
+
+void Config::CandidateSearch::Read(const boost::property_tree::ptree& params) {
+  ReadParamOptional(search_radius, params, "default.search_radius");
+  CHECK_NONNEGATIVE_PARAM(search_radius, "search_radius");
+
+  ReadParamOptional(max_search_radius, params, "default.max_search_radius");
+  CHECK_POSITIVE_PARAM(max_search_radius, "max_search_radius");
+
+  ReadParamOptional(cache_size, params, "grid.cache_size");
+  ReadParamOptional(grid_size, params, "grid.size");
+}
+
+void Config::Transition::Read(const boost::property_tree::ptree& params) {
+  ReadParamOptional(beta, params, "default.beta");
+  CHECK_POSITIVE_PARAM(beta, "beta");
+
+  ReadParamOptional(breakage_distance, params, "default.breakage_distance");
+  CHECK_POSITIVE_PARAM(breakage_distance, "breakage_distance");
+
+  ReadParamOptional(max_route_distance_factor, params, "default.max_route_distance_factor");
+  CHECK_POSITIVE_PARAM(max_route_distance_factor, "max_route_distance_factor");
+
+  ReadParamOptional(max_route_time_factor, params, "default.max_route_time_factor");
+  CHECK_POSITIVE_PARAM(max_route_time_factor, "max_route_time_factor");
+
+  ReadParamOptional(turn_penalty_factor, params, "default.turn_penalty_factor");
+  CHECK_NONNEGATIVE_PARAM(turn_penalty_factor, "turn_penalty_factor");
+}
+
+void Config::Emission::Read(const boost::property_tree::ptree& params) {
+  ReadParamOptional(sigma_z, params, "default.sigma_z");
+  CHECK_POSITIVE_PARAM(sigma_z, "sigma_z");
+
+  ReadParamOptional(gps_accuracy, params, "default.gps_accuracy");
+  CHECK_NONNEGATIVE_PARAM(gps_accuracy, "gps_accuracy");
+}
+
+void Config::Routing::Read(const boost::property_tree::ptree& params) {
+  ReadParamOptional(interpolation_distance, params, "default.interpolation_distance");
+  CHECK_POSITIVE_PARAM(interpolation_distance, "interpolation_distance");
+}
+
+void Config::Customizable::Reset() {
+  turn_penalty_factor = false;
+  search_radius = false;
+  gps_accuracy = false;
+  breakage_distance = false;
+  interpolation_distance = false;
+}
+
+void Config::Customizable::Read(const boost::property_tree::ptree& params) {
```

Comment:
See comment below about possibly moving the flag into each struct. This property of each configuration should be colocated with each value its describing.